slider/beatmap.py

```diff
@@ -257,12 +257,22 @@ class HitObject:
         Unknown currently.
     """
     time_related_attributes = frozenset({'time'})
-
-    def __init__(self, position, time, hitsound, addition='0:0:0:0:'):
+    type_code = 0
```

Comment:
hmm...it's already hard for consumers to subclass `HitObject` in a useful way as class references are hardcoded in `HitObject#parse`, so I'm not sure how much adding this attribute this matters for consumers when it's already set for all internal members. I guess it's not hurting anyone, though.

I have however changed it to `None`, as `type_code = 0` is likely to silently succeed in places an int is expected and break things if it ever falls through to a subclass.

Reply:
I specifically made it an integer so `_get_type_bits()` wouldn't throw a `TypeError` when its trying to use the `|` operator on a NoneType and int

Reply:
well, my point is that method throwing a type error when `type_code` isn't set is good. That's what I meant by things "silently breaking" - if we ever subclass `HitObject` and forget to set a `type_code`, that should show up as a failure, not silently "work" (but produce incorrect results).

Reply:
yes i realized that now